### ISSUE TITLE
How to get a  fine-tuned image-caption huggingface-version OFA model ?

### ISSUE THREAD

**Issue by PhoebusSi** [OFA-Sys/OFA#171]
How could I get the fined-tuned image-caption OFA model of **huggingface version**, which had topped the MSCOCO Image Caption Leaderboard ?

**Comment by JustinLin610** [comment]
Sorry for my late response. To reach a similar score on the leaderboard, you should use the finetuned checkpoint. We have already released the huge ckpt of captioning for the original code, and we'll soon release the one for HF transformers, I guess this week.

**Comment by steve-marmalade** [comment]
Is it possible to load the finetuned checkpoints using the huggingface model definition (e.g. `from transformers import OFAModel`)? The [uploaded files](https://huggingface.co/OFA-Sys/OFA-base-caption/tree/main) only include the model checkpoint, not the model config. I tried copying over the config from `OFA-base`, but got some warnings about uninitialized layers and the generated captions are gibberish. Thanks!

**Comment by NohTow** [comment]
No, the model definition is not the same as HF's, you basically start from a randomly initialised network.

**Comment by monatis** [comment]
Hi @JustinLin610, congrats on the awesome work.

> and we'll soon release the one for HF transformers, 

Do you have any any update in this?

@NohTow:

> No, the model definition is not the same as HF's, 

So does it mean that they are artifacts of uncompatible architectures, or is it possible to map parameter names from the fairseq version to HF Transformers version?

I'd like to help with it if it's possible.

**Comment by NohTow** [comment]
I think it should be possible to translate the fairseq version to an HF one (since I guess that is what they did for the other regular versions) but I am not sure nor I know how to do this.

**Comment by monatis** [comment]
Ok I quickly had a look at it and for the base-sized checkpoint ~100 parameters out of 965 have different names. I'll try to make a conversion tomorrow --it's definitely not a job for Friday night 😬

**Comment by monatis** [comment]
I managed to convert weights from Fairseq version to Transformers-compatible one. Here's a [PR on HF Hub](https://huggingface.co/OFA-Sys/OFA-base-caption/discussions/1) for the base version --I'm also making another PR for the large size shortly.

@JustinLin610

**Comment by monatis** [comment]
There's no HF model repo for OFA-large-caption, so I couldn't make a PR. Instead, I uploaded the converted model to [my storage for public download](https://storage.googleapis.com/mys-released-models/OFA-large-caption.zip).

**Comment by NohTow** [comment]
Very cool, thank you very much !
I’ll test it on monday and try to reproduce the paper results.

Converting models weights always seems a bit tricky but doable, do you mind sharing some ressources/insights on how you did this ? It has to do with named parameters, but if you have any useful inputs on that, it would be awesome !

But again, thank you already for doing it for this model !

**Comment by monatis** [comment]
> I’ll test it on monday and try to reproduce the paper results.

Cool. I'd like to hear about the results of your tests as well. I'll also share the converted huge-sized captioning model --it's the model that topped the COCO Learderboard. 

> if you have any useful inputs on that, it would be awesome 

Sure. I'll share my code with some explanatory notes on Monday.

**Comment by monatis** [comment]
And here comes the huge-sized image-captioning model. [Download, zipped, 2.39 GB](https://storage.googleapis.com/mys-released-models/OFA-huge-caption.zip).

**Comment by monatis** [comment]
And here's the [Colab notebook with explanations](https://colab.research.google.com/drive/1LLJewY92LXdeug5m_ceMUHdlqrRQwSQJ?usp=sharing) that I used for conversion.

**Comment by NohTow** [comment]
> And here's the [Colab notebook with explanations](https://colab.research.google.com/drive/1LLJewY92LXdeug5m_ceMUHdlqrRQwSQJ?usp=sharing) that I used for conversion.

Thank you very much, I'll bookmark it so I know how to do this kind of translation in the future !




> I managed to convert weights from Fairseq version to Transformers-compatible one. Here's a [PR on HF Hub](https://huggingface.co/OFA-Sys/OFA-base-caption/discussions/1) for the base version --I'm also making another PR for the large size shortly.

So I tried using these weights (each version, from base to huge) in my own loop based on HF examples and it seems like the model is correctly loaded: 
```
[INFO|modeling_utils.py:1703] 2022-09-12 14:51:32,183 >> All model checkpoint weights were used when initializing OFAModel.

[INFO|modeling_utils.py:1711] 2022-09-12 14:51:32,183 >> All the weights of OFAModel were initialized from the model checkpoint at /nfs/nas4.irisa.fr/deepfakes/deepfakes/image_repurposing/multimodal-fake-news-detection/VisualNews/OFA/weights/OFA-base-caption.
If your task is similar to the task the model of the checkpoint was trained on, you can already use OFAModel for predictions without further training.
```
Also, the resulting model seems to be able to generate ok-ish captions, so I guess we can confirm that weights are indeed well imported (would result in total gibberish otherwise).

For example, it generates `"<s> an airplane is parked on the runway at an airport}})\\)\\)\\)\\)\\)\\</s>"` when one reference is `"the airplane has landed behind a fence with barbed wire"` so the model definitely works as intended. Yet, at you might have noticed, at the end of the caption, it often (nearly always ?) generate gibberish before the eos token.

I remember this issue https://github.com/OFA-Sys/OFA/issues/172 that seems very related. However, it should not happened on an "official version" of the model, so it's weird.

**Comment by NohTow** [comment]
> I observed this phenomenon in various other Seq2Seq models (i.e., jibberish artifact at the end of the sequence), and it's easy to get rid of them, so I didn't bother

Yeah sure we can filter it out, but it still means that the model tries to output this, which is not really a good behavior and I guess it is hurting the performance in the end. 


> The issue mentions about a fix for the captioning task, so it might be that the captioning checkpoint is exported before this fix. But this is only a guess --maybe we can a confirmation from the authors.

Yes, but again, it would be really strange that a model that output gibberish in the end will be SOTA imho. Maybe the filtering is enough to get good results and the learning is not hurt but I find it odd.



> In their HF Spaces app, the author set that value to `cfg.task.patch_image_size` (here on [line 53](https://huggingface.co/spaces/OFA-Sys/OFA-Image_Caption/blob/main/app.py#L53)), and it's defined as 480 by default here on [line 62](https://github.com/OFA-Sys/OFA/blob/main/tasks/ofa_task.py#L62).
> Using a resolution of 480 makes the inference run longer, but a lower resolution lead to missing some details such as small or rare objects and colors. So I chose to stick to the value of 480 because it seems to yield more elaborate captions.

Very well, but maybe it would be nice to know the exact experimental setup if we want to reproduce the results.
I need to double check, but should not we use the resolution it has been trained on or it does not matter there ?

Overall, it would be really nice to have every information needed to reproduce the results with HF versions. I guess we can find them in evaluation and training scripts used with fairseq, but it needs some digging.

**Comment by monatis** [comment]
> I guess it is hurting the performance in the end

Don't think so. As previously stated, most of the Seq2Seq models have this behavior. I also observed it in my TTS research. A high-quality TTS model I trained always produced one-second noise at the end of the synthesized speech, and I was simply trimming it out.

I also verified that the authors [use postprocessing to remove them](https://github.com/OFA-Sys/OFA/blob/main/utils/eval_utils.py#L36).

----------

I observed something interesting with following image:
![boxing](https://user-images.githubusercontent.com/18634956/189909362-d1922688-acab-4f14-846a-47e190e28f0e.jpeg)

HD model produced the following with `use_cache=True`: "a man is standing next to a red punching bag".
However, it produced the following more detailed caption with `use_cache=False`: "a man wearing a helmet and boxing gloves standing next to a red punching bag".
Interestingly, the beam search decoder in Fairseq produced exactly the latter with `use_cache=True`, so it's faster and more elaborate. Again, this has nothing to do with the model itself, but it's related to how the beam search decoding is implemented. For some reason, we need to set `use_cache=False` in HF to match the quality of the Fairseq beam search decoding, thus ending up with 3-to-4x slower inference.

**Comment by NohTow** [comment]
> Don't think so. As previously stated, most of the Seq2Seq models have this behavior. I also observed it in my TTS research. A high-quality TTS model I trained always produced one-second noise at the end of the synthesized speech, and I was simply trimming it out.
> 
> I also verified that the authors [use postprocessing to remove them](https://github.com/OFA-Sys/OFA/blob/main/utils/eval_utils.py#L36).

Well, I evaluated the model with your regex to trim the bad ending and got these results:
```
- base-caption : 'eval_bleu': 0.3787352658847889
- large-caption : 'eval_bleu': 0.4061869408828353,
- huge-caption : eval_bleu': 0.4294927572391054
```
So results seems ok-ish so yeah it should do the trick, but they are still a bit different from reported ones. I'll try with the trimming used by the authors and double check that I have the exact same data as the paper (as well as my pipeline).

> nterestingly, the beam search decoder in Fairseq produced exactly the latter with `use_cache=True`, so it's faster and more elaborate. Again, this has nothing to do with the model itself, but it's related to how the beam search decoding is implemented. For some reason, we need to set `use_cache=False` in HF to match the quality of the Fairseq beam search decoding, thus ending up with 3-to-4x slower inference.

Very interesting. I'll make sure to use `use_cache=False` for my attempt to reproduce results. Are you sure that this is due to how beam search is implemented or it might be coming from something wrong in the cached hidden states ? I mean, is it worth it for me to dig into this and debug or it is the expected behavior ?

Also, do you know which decoding method/parameters are used in the paper for image captioning ? It is mentioned for NLU but I can't find it for captioning.

**Comment by yangjianxin1** [comment]
Maybe you can try the repository,I succeed to use transformers to train OFA model and inference: https://github.com/yangjianxin1/OFA-Chinese

**Comment by ghost** [comment]
Anyone have checkpoints for the tiny model? My use-case is doing basic image captioning on jpg images.

Specifically though I have only two use cases for local or remote -
Image captionaing using OFA-tiny 
Question answering using OFA-tiny

IF the checkpoints, even just two per model were provided, one for FP32 and one for FP16 or mixed FP, the community can do the rest. But without a base to build on, this entire paper, entire arch and this repo will be forgotten.

Also in the model page for OFA-tiny they mention import but this currently will never work -

```
from generate import sequence_generator
```

PS - I can see that the last comment here was in 2023. And transformers still does not support the OFA models or the arch, not even the tokenizer is supported.

Manish
Check out my kind of oldfangled, kinda newfangled, kinda cool, inevitably thought-provoking profile. And the limited mentorship opportunities for your no matter the non business major you have chosen be it - Bachelor's or Master's or Doctorates or Post-Doctorates.
